core/src/server/rpc_module.rs

```diff
@@ -948,21 +968,22 @@ impl SubscriptionSink {
 	///     tokio::spawn(async move { sink.pipe_from_stream(stream).await; });
 	/// });
 	/// ```
-	pub async fn pipe_from_stream<S, T>(&mut self, stream: S) -> SubscriptionClosed
+	pub async fn pipe_from_stream<S, T, F>(self, stream: S, on_close: F)
 	where
 		S: Stream<Item = T> + Unpin,
 		T: Serialize,
+		F: FnOnce(SubscriptionClosed, SubscriptionSink),
 	{
-		self.pipe_from_try_stream::<_, _, Error>(stream.map(|item| Ok(item))).await
+		self.pipe_from_try_stream::<_, _, Error, _>(stream.map(|item| Ok(item)), on_close).await
 	}
 
-	/// Returns whether this channel is closed without needing a context.
+	/// Returns whether the subscription is closed.
 	pub fn is_closed(&self) -> bool {
-		self.inner.is_closed() || self.close_notify.is_none()
+		self.inner.is_closed() || self.close_notify.is_none() || self.is_active_subscription()
 	}
 
 	fn is_active_subscription(&self) -> bool {
-		Arc::strong_count(&self.active_sub) > 1
+		self.subscribers.lock().contains_key(&self.uniq_sub)
```

Comment:
this is a bit annoying with extra mutex accesses but I kept for ease of use.

we need another `AtomicBool` or something because `rx side` of oneshot can't be checked